### ISSUE TITLE
Multiple COLLECT gives wrong results with 3.8, works in 3.7.10

### ISSUE THREAD

**Issue by matcho** [arangodb/arangodb#14672]
## My Environment

* __ArangoDB Version__:        3.8.0
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Leader/Follower ("Master/Slave")
* __Deployment Strategy__:   systemd
* __Configuration__:               Leader 3.7.10 replicated to follower 3.8.0
* __Infrastructure__:               own
* __Operating System__:        Ubuntu 18.04 (L) and 20.04 (F)
* __Total RAM in your machine__:        256GB (L) and 32 GB (F)
* __Disks in use__:        HDD
* __Used Package__:              Ubuntu .deb

## Component, Query & Data

__Affected feature__:
AQL query, COLLECT statement.

__AQL query (if applicable)__:
```
FOR q IN queries
    SORT q.date DESC
    LIMIT 1000
    COLLECT resultStatus = q.result.status, requestStatus = q.request.status WITH COUNT INTO nb
    RETURN { resultStatus, requestStatus, nb }
```
__AQL explain and/or profile (if applicable)__:
### Explain on 3.8.0
```
Query String (191 chars, cacheable: true):
 FOR q IN queries
     SORT q.date DESC
     LIMIT 1000
     COLLECT resultStatus = q.result.status, requestStatus = q.request.status WITH COUNT INTO nb 
 RETURN { resultStatus, requestStatus, nb }

Execution plan:
 Id   NodeType               Est.   Comment
  1   SingletonNode             1   * ROOT
 12   IndexNode         450882273     - FOR q IN queries   /* reverse skiplist index scan, projections: `request` */    
  5   LimitNode              1000       - LIMIT 0, 1000
  6   CalculationNode        1000       - LET #6 = q.`result`.`status`   /* attribute expression */   /* collections used: q : queries */
  7   CalculationNode        1000       - LET #8 = q.`request`.`status`   /* attribute expression */   /* collections used: q : queries */
  8   CollectNode             800       - COLLECT resultStatus = #6, requestStatus = #8 AGGREGATE nb = LENGTH()   /* hash */
 11   SortNode                800       - SORT resultStatus ASC, requestStatus ASC   /* sorting strategy: standard */
  9   CalculationNode         800       - LET #10 = { "resultStatus" : resultStatus, "requestStatus" : requestStatus, "nb" : nb }   /* simple expression */
 10   ReturnNode              800       - RETURN #10

Indexes used:
 By   Name        Type       Collection   Unique   Sparse   Selectivity   Fields       Ranges
 12   idx_49159   skiplist   queries      false    false        99.64 %   [ `date` ]   *

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-calculations-up-2
  3   use-indexes
  4   use-index-for-sort
  5   remove-unnecessary-calculations-2
  6   move-calculations-down
  7   reduce-extraction-to-projection

Optimization rules with highest execution times:
 RuleName                                    Duration [s]
 use-indexes                                      0.00009
 specialize-collect                               0.00003
 reduce-extraction-to-projection                  0.00002
 remove-unnecessary-calculations-2                0.00001
 use-index-for-sort                               0.00001

76 rule(s) executed, 2 plan(s) created
```

### Explain on 3.7.10
```
Query String (191 chars, cacheable: true):
 FOR q IN queries
     SORT q.date DESC
     LIMIT 1000
     COLLECT resultStatus = q.result.status, requestStatus = q.request.status WITH COUNT INTO nb 
 RETURN { resultStatus, requestStatus, nb }

Execution plan:
 Id   NodeType               Est.   Comment
  1   SingletonNode             1   * ROOT
 12   IndexNode         450916894     - FOR q IN queries   /* reverse skiplist index scan, projections: `request`, `result` */    
  5   LimitNode              1000       - LIMIT 0, 1000
  6   CalculationNode        1000       - LET #6 = q.`result`.`status`   /* attribute expression */   /* collections used: q : queries */
  7   CalculationNode        1000       - LET #8 = q.`request`.`status`   /* attribute expression */   /* collections used: q : queries */
  8   CollectNode             800       - COLLECT resultStatus = #6, requestStatus = #8 WITH COUNT INTO nb   /* hash */
 11   SortNode                800       - SORT resultStatus ASC, requestStatus ASC   /* sorting strategy: standard */
  9   CalculationNode         800       - LET #10 = { "resultStatus" : resultStatus, "requestStatus" : requestStatus, "nb" : nb }   /* simple expression */
 10   ReturnNode              800       - RETURN #10

Indexes used:
 By   Name        Type       Collection   Unique   Sparse   Selectivity   Fields       Ranges
 12   idx_49159   skiplist   queries      false    false        99.43 %   [ `date` ]   *

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-calculations-up-2
  3   use-indexes
  4   use-index-for-sort
  5   remove-unnecessary-calculations-2
  6   move-calculations-down
  7   reduce-extraction-to-projection
```

__Dataset__:
`queries` collection has 450M docs. In the 1000 docs selected in the query above, `q.request.status` is always `null`.

__Size of your Dataset on disk__:
A few hundred GBs

## Steps to reproduce

1. Run query above on 3.7.10. Result, as expected, is

| resultStatus | requestStatus | nb |
| --- | --- | --- |
| null | null | 1 |
| ok | null | 893 |
| reject | null | 22 |
| wrong_referential | null | 84 |

2. Run same query on 3.8.0, result is

resultStatus | requestStatus | nb
-- | -- | --
null | null | 1000

3. Edit the query to collect only one column on 3.8
```
FOR q IN queries
    SORT q.date DESC
    LIMIT 1000
    COLLECT resultStatus = q.result.status WITH COUNT INTO nb RETURN { resultStatus, nb }
```
and result is as expected

resultStatus | nb
-- | --
null | 1
ok | 893
reject | 22
wrong_referential | 84

__Problem__:
Multiple collect seems to break results in 3.8.0.
Is it because of `q.request.status` being always `null` in this subset of data ? Or because an index is used to retrieve subset of documents ? Rules applied seem quite different in respective explains.

Thanks a lot !

__Expected result__:
Same result as in 3.7.10.

**Comment by Simran-B** [comment]
Thanks for reporting this! I was able to narrow it down to a problem with the `reduce-extraction-to-projection` optimizer rule and nested attribute access with the same attribute name at the same level (here: `status`):

```js
db._drop("coll");
db._create("coll");
db.coll.save({ result: {} });
db.coll.save({ result: {status: "ok" } });

/* wrong with result.status / request.status (both "status") */
db._query(`FOR d IN coll COLLECT resultStatus = d.result.status, requestStatus = d.request.status WITH COUNT INTO count RETURN { resultStatus, requestStatus, count }`);

/* fine with optimizer rule disabled */
db._query(`FOR d IN coll COLLECT resultStatus = d.result.status, requestStatus = d.request.status WITH COUNT INTO count RETURN { resultStatus, requestStatus, count }`,
  null, { optimizer: { rules: ["-reduce-extraction-to-projection"] }});

/* fine with result.status / request.other (different nested attribute names) */
db._query(`FOR d IN coll COLLECT resultStatus = d.result.status, requestOther = d.request.other WITH COUNT INTO count RETURN { resultStatus, requestOther, count }`);
```

I created an internal ticket BTS-562.

**Comment by jsteemann** [comment]
Fixed via https://github.com/arangodb/arangodb/pull/14801

**Comment by matcho** [comment]
Hi. We can confirm that the bug is fixed in v3.8.3

Many thanks !
